test/unit/test-builder.js

```diff
@@ -0,0 +1,422 @@
+import {READY_SCAN_SIGNAL} from '../../src/service/resources-interface';
+import {createElementWithAttributes} from '../../src/dom';
+import {installIntersectionObserverStub} from '../../testing/intersection-observer-stub';
+
+describes.realWin('Builder', {amp: true}, (env) => {
+  let win, doc, ampdoc;
+  let setAmpdocReady;
+  let clock;
+  let intersectionObserverStub;
+  let builder;
+
+  beforeEach(() => {
+    win = env.win;
+    doc = win.document;
+    ampdoc = env.ampdoc;
+
+    let ampdocReady = false;
+    let ampdocReadyResolver;
+    const ampdocReadyPromise = new Promise((resolve) => {
+      ampdocReadyResolver = resolve;
+    });
+    setAmpdocReady = () => {
+      ampdocReady = true;
+      ampdocReadyResolver();
+      return ampdocReadyPromise.then(() => {});
+    };
+    env.sandbox.stub(ampdoc, 'whenReady').returns(ampdocReadyPromise);
+    env.sandbox.stub(ampdoc, 'isReady').callsFake(() => ampdocReady);
+
+    delete win.requestIdleCallback;
+    clock = fakeTimers.withGlobal(win).install();
+    win.requestIdleCallback = (callback) => {
+      win.setTimeout(callback, 100);
+    };
+
+    intersectionObserverStub = installIntersectionObserverStub(
+      env.sandbox,
+      win
+    );
+
+    builder = new Builder(ampdoc);
+  });
+
+  afterEach(() => {
+    clock.uninstall();
+  });
+
+  function createAmpElement(options = {}) {
+    const element = createElementWithAttributes(doc, 'amp-el', {});
+    element.deferredBuild = () => options.deferredBuild || false;
+    element.prerenderAllowed = () => options.prerenderAllowed || false;
+    element.getBuildPriority = () =>
+      options.buildPriority || LayoutPriority.CONTENT;
+    element.mutable = () => options.mutable || false;
+    element.buildInternal = env.sandbox.stub();
+    return element;
+  }
+
+  describe('schedule', () => {
+    it('should schedule a deferredBuild element', () => {
+      const element = createAmpElement({deferredBuild: true});
+      builder.schedule(element);
+      expect(intersectionObserverStub.isObserved(element)).to.be.true;
+
+      builder.unschedule(element);
+      expect(intersectionObserverStub.isObserved(element)).to.be.false;
+    });
+
+    it('should schedule a non-deferredBuild element', () => {
+      const element = createAmpElement({deferredBuild: false});
+      builder.schedule(element);
+      expect(intersectionObserverStub.isObserved(element)).to.be.false;
+    });
+
+    it('should unschedule when built', async () => {
+      const element = createAmpElement({deferredBuild: true});
+      builder.schedule(element);
+      expect(intersectionObserverStub.isObserved(element)).to.be.true;
+
+      await setAmpdocReady();
+      intersectionObserverStub.notifySync({
+        target: element,
+        isIntersecting: true,
+      });
+      expect(intersectionObserverStub.isObserved(element)).to.be.false;
+    });
+
+    it('should signal READY_SCAN_SIGNAL after first element scheduled', async () => {
```

Comment:
FMI: I couldn't figure out what READY_SCAN_SIGNAL is used for. I see its used by `whenContentIniLoadInOb`, but would we be better served by waiting for [DOMContentLoaded](https://developer.mozilla.org/en-US/docs/Web/API/Window/DOMContentLoaded_event) instead?

Reply:
I created a compatible version of `READY_SCAN_SIGNAL` with the `Resources` class. I feel like it's certainly could be done differently and possibly better. AFAIK it's only used for `whenConentIniLoad` (but both versions: old and new). It doesn't correlate very well with `DOMContentLoaded` however. The goal of this signal is to pick up N early elements once they start upgrading and building. `DOMContentLoaded` can be both too early and too late. And to stress: here I just tried to reproduce as close as possible whatever the `Resources` class has been doing.

Reply:
Gotchya. I'll mark a personal TODO item then to try figure out what this is used for and ensure we are creating a meaningful result.

Reply:
Let's dedupe to https://github.com/ampproject/amphtml/pull/32568#discussion_r578760097